aesd-char-driver/main.c

```diff
@@ -48,6 +49,134 @@ int aesd_release(struct inode *inode, struct file *filp)
     return 0;
 }
 
+loff_t aesd_llseek(struct file *filp, loff_t offset, int whence) 
+{
+    struct aesd_dev* dev = filp->private_data;
+    struct aesd_circular_buffer* cbuf = dev->cbuf;
+    
+    loff_t size = 0; //set to size of full circular buffer
+    
+    uint8_t index;
+    struct aesd_buffer_entry *entry;
+    mutex_lock(dev->lock_cc);
+    AESD_CIRCULAR_BUFFER_FOREACH(entry,cbuf,index) {
+        size = size + entry->size;
+    }
+    mutex_unlock(dev->lock_cc);
+    
+    loff_t new_pos = fixed_size_llseek(filp, offset, whence, size);
+    
+    return new_pos;
+}
+
+/**
+ * Adjust the file offset (f_pos) parameter of @param filp based on the location specified by
+ * @param write_cmd (the zero referenced command to locate)
+ * and @param write_cmd_offset (the zero referenced offset into the command)
+ * @return 0 if successful, negative if error occurred:
+ *      -ERESTARTSYS if mutex could not be obtained
+ *      -EINVAL if write command or wrtie_cmd_offset was out of range
+ */
+static long aesd_adjust_file_offset(struct file* filp, unsigned int write_cmd, 
+					unsigned int write_cmd_offset)
+{
+    PDEBUG("ioctl: cmd=%d and offset=%d", write_cmd, write_cmd_offset);
+    long retval = 0;
+    
+    struct aesd_dev* dev = filp->private_data;
+    struct aesd_circular_buffer* cbuf = dev->cbuf;
+    
+    //check for valid cmd
+    if(write_cmd >= AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED) 
+    {
+        retval = -EINVAL;
+        goto end;
+    }
+    //locking
+    mutex_lock(dev->lock_cc);
+    uint8_t out_o = cbuf->out_offs;
+    uint8_t in_o = cbuf->in_offs;
+    bool full = cbuf->full;
+    mutex_unlock(dev->lock_cc);
+    
+    uint8_t bufs = in_o - out_o;
+    if(full) bufs = AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED;
+    if(write_cmd > bufs) {
+        retval = -EINVAL;
+        goto end;
+    }
+    
+    //check for valid offset
+    uint8_t cmd_index = out_o + write_cmd;
+    if(cmd_index >= AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED) {
+        cmd_index = cmd_index - AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED;
+    }
+    
+    uint8_t cmd_size = cbuf->entry[cmd_index].size;
+    if(write_cmd_offset > cmd_size) {
+        retval = -EINVAL;
+        goto end;
+    }
+    
+    //calculate start offset to write_cmd
+    struct aesd_buffer_entry *entry;
+    uint8_t i = out_o;
+    size_t new_offs = 0;
+    while(i != cmd_index) {
+    	//set entry
+    	entry = &(cbuf->entry[i]);
+        
+        new_offs = new_offs + entry->size;
+        
+        //handle wrap-around
+        if(i == (AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED - 1))
+            i = 0;
+        else 
+            i++;
+    }
+    
+    //add write_cmd_offset
+    new_offs = new_offs + write_cmd_offset;
+    
+    //save output to filp->f_pos
+    mutex_lock(dev->lock_fpos);
+    filp->f_pos = new_offs;
```

Comment:
@mamo6538 this is not something you should lock, see discussion in class about this.